### ISSUE TITLE
Pixel coordinates clipping behaviour in gammapy.maps 

### ISSUE THREAD

**Issue by adonath** [gammapy/gammapy#1440]
When transforming from pixel coordinates to indices using the `MapGeom` object, the pixel coordinates are clipped, when outside of the image range and an index of `-1` is returned. This might be the desired behaviour in some cases but I found it surprising. The option `clip=False` is without effect on the result.

Here is a minimal example:
```python
from gammapy.maps import Map
from astropy.coordinates import SkyCoord

position = SkyCoord(0.0, 5.0, frame='galactic', unit='deg')

# Create a 100x100 pix WCS Map
m_wcs = Map.create(binsz=0.1, map_type='wcs', skydir=position, width=10.0)

# transform pixel coordinates to indices
m_wcs.geom.pix_to_idx((102, 102), clip=True)

# transform pixel coordinates to indices without clipping
m_wcs.geom.pix_to_idx((-23.3, -11.23), clip=False)
```
Which returns `(array([-1]), array([-1]))` in both cases with the option set to `clip=True` as well as `clip=False`. So there are actually two issues here:

* The option `clip=True/False` seems non-functional.
* Decide on a reasonable default behaviour (clipping vs. raising `IndexError` vs. just transform and raise the `IndexError` when the transformed index is actually used...)

**Comment by cdeil** [comment]
@adonath - See #1186 and #1208 where this was introduced.

As mentioned there, the idea is to return and "invalid" sentinel value instead of raising an error. This is what `astropy.wcs` does also. It's very common that one has arrays of coordinates, and wants the result for the good ones, even if some are invalid (e.g. in an AIT all-sky plot.

We should definitely introduce a constant for the sentinel value `-1` to make it easy to find all places where it is put and used. And then it also sounds like there's a bug. Maybe we can have a look together at the codebase and try to figure it out?

**Comment by adonath** [comment]
Yes, I think the minimal action would be to introduce a global constant and document the behaviour. What I found surprising is that `-1` is actually a valid index and is also returned for negative pixel coordinates, where I would rather expect that the values are clipped at `0`. But now I understand it's actually only used as a place-holder.

**Comment by cdeil** [comment]
The problem is that with `int` there is no good placeholder. If you want to make it a bit more obvious, you could use `-42`.

**Comment by cdeil** [comment]
@adonath - I'd like to release Gammapy v0.8 in a week. What should we do here? OK to defer to v0.9, or do you think this needs to be addressed now?

**Comment by cdeil** [comment]
I think we should try and just change to always use float where `np.nan` is available (see #2374), and only if needed go to `int` on I/O for counts maps. I'm moving this to v0.15 because we shouldn't do last-minute design changes in v1.0.

**Comment by adonath** [comment]
This is still an open issue but could be address as part of refactoring `gammapy.maps`

**Comment by registerrier** [comment]
Hi @adonath , what remains to be done here?

**Comment by registerrier** [comment]
This is further discussed in #6430 . Closing.
